File: C++/Medium/Accepted/magical-string_5.cpp
class Solution {
public:
    int magicalString(int n) {
        
        if (n == 0) return 0;
        if (n <= 3) return 1;
        vector<int> s(n + 1, 0);
        s[0] = 1;
        s[1] = 2;
        s[2] = 2;
        int i = 2;
        int j = 3;
        int num = 1;
        int ans = 1;
        while (j < n) {
            for (int k = 0; k < s[i]; k++) {
                s[j] = num;
                if (num == 1 && j < n) ans++;
                j++;
            }
            num = 3 - num;
            i++;
        }
        return ans;
    }
};

File: Rust/Medium/WrongAnswer/magical-string_5.rs
impl Solution {
    pub fn magical_string(n: i32) -> i32 {
        

        let mut s = vec![1, 2, 2];
        let mut i = 2;
        let mut count = 1;
        while s.len() < n as usize {
            let v = s[i];
            let last = s[s.len() - 1];
            if last == 1 {
                s.push(2);
            } else {
                s.push(1);
                count += 1;
            }
            if v == 2 {
                if last == 1 {
                    s.push(2);
                } else {
                    s.push(1);
                    count += 1;
                }
            }
            i += 1;
        }
        if s.len() == n as usize {
            count
        } else {
            count - (s[s.len() - 1] - (n as usize - s.len() + 1) as i32)
        }
    }
}